skbio/sequence/_grammared_sequence.py

```diff
@@ -22,11 +22,79 @@
 from ._sequence import Sequence
 
 
-class IUPACSequence(with_metaclass(ABCMeta, Sequence)):
-    """Store biological sequence data conforming to the IUPAC character set.
+class GrammaredSequenceException(TypeError):
+    pass
+
+
+class GrammaredSequenceMeta(ABCMeta, type):
+    def __new__(mcs, name, bases, dct):
+        cls = super(GrammaredSequenceMeta, mcs).__new__(mcs, name, bases, dct)
+
+        # Only perform validation on classes that aren't abstract.
+        if (type(cls.default_gap_char) is not abstractproperty and
```

Comment:
I'm confused about the current checks: there's a top-level guard for `gap_chars` and `default_gap_char`, and the validations only have the chance to run if those are present. However, there are guarded validations that could be performed even if `gap_chars` and `default_gap_char` are abstract (e.g., checks dealing with degenerate/nondegenerate characters). Can you remove the hierarchy and instead perform independent validations where possible? Defining a few bools at the top may be helpful because you're going to have that code duplicated quite a bit (maybe a better name for the bools is `concrete_default_gap_char`, `concrete_gap_chars`, etc....?)